src/test/resources/imp.kore

```diff
@@ -69,45 +235,311 @@ module IMP
+   * https://github.com/kframework/kore/pull/101
+   */
+
+  /* See Matching Logic paper, Section 9.2.
+   * bind, emp, and keys are functional symbols.
+   * merge is partial function:
+   * forall m1 m2 . merge(m1, m2) = bottom
+   *                \/ exists t . merge(m1, m2) = t
+   * merging two binds with the same key gets bottom
+   * merge(bind(X,Y),bind(X,Z)) = bottom
+   */
+
+endmodule []
+
+module ID
+
+  import INT []
+
+  sort Id{} []
+
+  symbol id{}(Int{}) : Id{} []
+
+  /* id is injective. */
+
+  axiom{}
+    \implies{Id{}}(
+      \and{Id{}}(id{}(I1:Int{}), id{}(I2:Int{})),
+      id{}(\and{Int{}}(I1:Int{}, I2:Int{})))
+  []
+
+endmodule []
 
-  symbol gamma0{S1,S2}(S1,S2) : Ctxt{S1,S2} []
+module CONTEXT
 
+  sort Ctxt{S1,S2} []
+
+  /* Context application are often written as _[_]. */
   symbol app{S1,S2}(Ctxt{S1,S2}, S1) : S2 []
 
-  alias gamma{S1,S2}(S1,S2) : Ctxt{S1,S2} [] // we could make it a symbol
+  symbol gamma{S1,S2}(S1,S2) : Ctxt{S1,S2} []
 
-  alias idctxt{S}() : Ctxt{S,S} []
+  /* app and gamma are functional symbols. */
 
-  axiom{S,R}
-    \equals{Ctxt{S,S},R}(
-      idctxt{S}(),
-      gamma{S,S}(X:S,X:S)) []
+  axiom{S1,S2,R}
+    \forall{R}(
+      C:Ctxt{S1,S2},
+      \forall{R}(
+        T:S1,
+        \exists{R}(
+          Z:S2,
+          \equals{S2,R}(
+            app{S1,S2}(C:Ctxt{S1,S2}, T:S1),
+            Z:S2))))
+  []
+
+  axiom{S1,S2,R}
+    \forall{R}(
+      HOLE:S1,
+      \forall{R}(
+        BODY:S2,
+        \exists{R}(
+          C:Ctxt{S1,S2},
+          \equals{Ctxt{S1,S2},R}(
+            gamma{S1,S2}(HOLE:S1, BODY:S2),
+            C:Ctxt{S1,S2}))))
+  []
+
+  /* Identity context. */
+  symbol idctxt{S}() : Ctxt{S,S} []
+
+  /** Axioms about contexts following the monad style.
+    * That is, there are two axioms about contexts as follows.
+    * (Unit) idctxt[T] = T
+    * (Multiplication) C1[C2[T]] = C[T] where C = gamma HOLE . C1[C2[HOLE]]
+    */
 
   axiom{S,R}
     \equals{S,R}(
-      T:S,
-      app{S,S}(idctxt{S}(), T:S)) []
+      app{S,S}(idctxt{S}(), T:S), T:S)
+  []
+
+  axiom{S,S1,S2,R}
+    \equals{S,R}(
+      app{S1,S}(
+        C1:Ctxt{S1,S},
+        app{S2,S1}(
+          C2:Ctxt{S2,S1},
+          T:S2)),
+      app{S2,S}(
+        gamma{S2,S}(
+          HOLE:S2,
+          app{S1,S}(
+            C1:Ctxt{S1,S},
+            app{S2,S1}(
+              C2:Ctxt{S2,S1},
+              HOLE:S2))),
+        T:S2))
+  []
+
+endmodule []
+
+/** The module of "~>" and "." in K.
+  */
+module ARROW
 
+  import CONTEXT []
 
-  // Start with IMP Specific Definitions
+  sort K{} []
+
+  /* The injection function from any sort S to K. */
+  symbol asK{S}(S) : K{} []
+
+  /* Two injectivity axioms:
+   *   not (asK{S1}(X:S1) /\ asK{S2}(Y:S2)) if S1 and S2 are different
+   *   asK{S}(X:S) /\ asK{S}(Y:S) -> asK{S}(X:S /\ Y:S)
+   */
+
+  axiom{S1,S2,#R}
+    \implies{#R}(
+      \not{#R}(
+        \equals{#Sort{},#R}(
+          #S1:#Sort{}, 

```

Comment:
You're relying here on S1 being lifted to the #S1 variable, which makes me feel a bit tensed because the connection is not obvious.

Maybe this is not a big issue that's worth fixing. However, a bigger problem is that there are wellformed patterns which are not liftable (i.e. their lifted form is not wellformed), e.g.
```
axiom{S, #R}
  \forall{#R}(
    #S:#Pattern{}
    \implies{K{}}(
      \and{K{}}(
        asK{S}(X:S),
        asK{S}(Y:S)),
      asK{S}(\and{S}(X:S, Y:S))))
```

Reply:
The fact that S1 is lifted to #S1:#Sort{}, is according to the Kore semantics (Section 9.2.3 of the Semantics of K document). The lift rule is the follows:
[[_]] : <object-sort> -> <meta-pattern>
[[s]] = '#' s ':' '#Sort{}'    if s is a sort variable.

Reply:
@virgil-serbanuta Could you please inform me where your example comes from? I can recognize that part of it is the axiom in Line 367, but why there's a \forall{#R} in front of it?

Reply:
The origin of this example is not important. Yes, indeed, I took one of your axioms and changed it in order to provide an illustration for the issue I mentioned, but that's not the point. The main question is: do we want to allow kore to have wellformed patterns which cannot be lifted?

If yes, we should change our code to do a wellformedness check after lifting, and handle lifting errors (I think we currently assume that lifting just works). If not, we should either change the wellformedness requirements for kore or we should change how lifting is done.